File: web/src/lib/format.ts
export const statusLabel: Record<string, string> = {
  SCHEDULED: 'Скоро',
  LIVE: 'Live',
  PAUSED: 'Пауза',
  FINISHED: 'Финал',
  CANCELLED: 'Отменён',
  DRAFT: 'Черновик',
  REGISTRATION: 'Набор',
  ACTIVE: 'Идёт',
  PENDING: 'На рассмотрении',
  APPROVED: 'В сетке',
  REJECTED: 'Отказ',
  WITHDRAWN: 'Снялись',
}

export const sportLabel: Record<string, string> = {
  FOOTBALL: 'Футбол',
  FUTSAL: 'Футзал',
  BASKETBALL: 'Баскетбол',
  VOLLEYBALL: 'Волейбол',
  HOCKEY: 'Хоккей',
}

export const formatLabel: Record<string, string> = {
  ROUND_ROBIN: 'Круговой турнир',
  CUP: 'Кубок / плей-офф',
  GROUPS_PLAYOFF: 'Группы + плей-офф',
  SWISS: 'Швейцарская система',
  DOUBLE_ELIMINATION: 'Double elimination',
}

export function isTournamentOpenForApply(status?: string | null) {
  return status === 'REGISTRATION' || status === 'DRAFT'
}

export function formatDateRange(start?: string | null, end?: string | null) {
  const from = formatIsoDate(start)
  const to = formatIsoDate(end)
  if (!from && !to) return ''
  if (from && to && from !== to) return `${from} — ${to}`
  return from || to
}

function formatIsoDate(value?: string | null) {
  if (!value) return ''
  const match = String(value).match(/^(\d{4})-(\d{2})-(\d{2})/)
  if (match) return `${match[3]}.${match[2]}.${match[1]}`
  const formatted = formatMatchDay(value)
  return formatted === '—' ? '' : formatted
}

export function labelOfSport(code?: string | null, fallback?: string | null) {
  if (!code) return fallback || 'Спорт'
  return sportLabel[code] || fallback || code
}

export const roleLabel: Record<string, string> = {
  FAN: 'Зритель',
  PLAYER: 'Игрок',
  CAPTAIN: 'Капитан',
  REFEREE: 'Судья',
  ADMIN: 'Админ',
}

export const eventLabel: Record<string, string> = {
  GOAL: 'Гол',
  ASSIST: 'Голевая',
  YELLOW_CARD: 'Жёлтая',
  RED_CARD: 'Красная',
  SUBSTITUTION: 'Замена',
  POINT: 'Очко',
  FOUL: 'Фол',
  PERIOD_START: 'Начало тайма',
  PERIOD_END: 'Конец тайма',
  OTHER: 'Событие',
}

export function labelOf(map: Record<string, string>, value?: string | null, fallback = '—') {
  if (!value) return fallback
  return map[value] ?? value
}

export function formatWhen(value?: string | number | Date | null) {
  if (!value) return '—'
  const date = new Date(value)
  if (Number.isNaN(date.getTime())) return '—'
  return date.toLocaleString('ru-RU', {
    day: 'numeric',
    month: 'short',
    hour: '2-digit',
    minute: '2-digit',
  })
}

export function ageLine(value?: string | number | Date | null) {
  if (!value) return ''
  const date = new Date(value)
  if (Number.isNaN(date.getTime())) return ''
  const now = new Date()
  let age = now.getFullYear() - date.getFullYear()
  const monthDiff = now.getMonth() - date.getMonth()
  if (monthDiff < 0 || (monthDiff === 0 && now.getDate() < date.getDate())) age -= 1
  const day = String(date.getDate()).padStart(2, '0')
  const month = String(date.getMonth() + 1).padStart(2, '0')
  return `Возр: ${age} (${day}.${month}.${date.getFullYear()})`
}

export function formatMatchDay(value?: string | number | Date | null) {
  if (!value) return '—'
  const date = new Date(value)
  if (Number.isNaN(date.getTime())) return '—'
  const day = String(date.getDate()).padStart(2, '0')
  const month = String(date.getMonth() + 1).padStart(2, '0')
  const year = String(date.getFullYear()).slice(-2)
  return `${day}.${month}.${year}`
}

export const outcomeMark: Record<string, string> = {
  WIN: 'В',
  DRAW: 'Н',
  LOSS: 'П',
}

export function formatClock(totalSeconds?: number | null) {
  const safe = Math.max(0, totalSeconds ?? 0)
  const minutes = Math.floor(safe / 60)
  const seconds = safe % 60
  return `${minutes}:${String(seconds).padStart(2, '0')}`
}

export function periodNoun(sportCode?: string | null) {
  if (sportCode === 'BASKETBALL') return 'четверть'
  if (sportCode === 'VOLLEYBALL') return 'партия'
  return 'тайм'
}

export function periodLabel(period?: number | null, sportCode?: string | null, periodCount?: number | null) {
  if (!period) return 'Ещё не свистнули'
  if (periodCount && period > periodCount) return 'Доп. время'
  const noun = periodNoun(sportCode)
  if (noun === 'четверть') return `${period}-я четверть`
  if (noun === 'партия') return `${period}-я партия`
  return `${period}-й тайм`
}

export function playerTag(name?: string | null, jersey?: number | null) {
  if (!name && jersey == null) return ''
  if (jersey == null) return name || ''
  return `#${jersey} ${name || ''}`.trim()
}

export function eventDetail(ev: {
  eventType?: string
  playerName?: string | null
  playerJersey?: number | null
  secondaryPlayerName?: string | null
  secondaryPlayerJersey?: number | null
}) {
  const main = playerTag(ev.playerName, ev.playerJersey)
  const second = playerTag(ev.secondaryPlayerName, ev.secondaryPlayerJersey)
  if (ev.eventType === 'GOAL' && second) return `${main} · пас ${second}`
  if (ev.eventType === 'SUBSTITUTION' && second) return `${main} → ${second}`
  return main
}

export function initials(name?: string | null) {
  if (!name) return 'SL'
  const parts = name.trim().split(/\s+/).slice(0, 2)
  return parts.map((part) => part[0]?.toUpperCase() ?? '').join('') || 'SL'
}

export function passwordHint(password: string) {
  if (!password) return 'Минимум 8 символов. Можно без спецэффектов — главное, чтобы помнили.'
  if (password.length < 8) return 'Ещё чуть-чуть: пароль любит длину больше, чем остроумие.'
  if (password.length < 12) return 'Уже можно. Если добавите ещё пару знаков — судья кивнёт.'
  return 'Крепко. Этот пароль не забьёт автогол.'
}

File: mobile/lib/core/format.dart
const roleLabels = {
  'FAN': 'Зритель',
  'PLAYER': 'Игрок',
  'CAPTAIN': 'Капитан',
  'REFEREE': 'Судья',
  'ADMIN': 'Админ',
};

String initials(String? name) {
  if (name == null || name.trim().isEmpty) return 'SL';
  final parts = name.trim().split(RegExp(r'\s+')).take(2);
  return parts.map((p) => p[0].toUpperCase()).join();
}

String ymd(DateTime date) {
  final m = date.month.toString().padLeft(2, '0');
  final d = date.day.toString().padLeft(2, '0');
  return '${date.year}-$m-$d';
}

String shortKickoff(DateTime? at, String status) {
  if (status == 'LIVE' || status == 'PAUSED') return 'LIVE';
  if (at == null) return '—';
  final d = at.toLocal();
  final day = d.day.toString().padLeft(2, '0');
  final month = d.month.toString().padLeft(2, '0');
  if (status == 'FINISHED' || status == 'CANCELLED') return '$day.$month.';
  final h = d.hour.toString().padLeft(2, '0');
  final min = d.minute.toString().padLeft(2, '0');
  return '$h:$min';
}

String longKickoff(DateTime? at) {
  if (at == null) return '—';
  final d = at.toLocal();
  final day = d.day.toString().padLeft(2, '0');
  final month = d.month.toString().padLeft(2, '0');
  final h = d.hour.toString().padLeft(2, '0');
  final min = d.minute.toString().padLeft(2, '0');
  return '$day.$month.${d.year} $h:$min';
}

String matchStateLabel(String status) {
  return switch (status) {
    'FINISHED' => 'Завершен',
    'LIVE' => 'Live',
    'PAUSED' => 'Пауза',
    'CANCELLED' => 'Отменён',
    'SCHEDULED' => 'Не начался',
    _ => status,
  };
}

int eventMinute(int? gameTime) {
  final raw = gameTime ?? 0;
  if (raw < 0) return 0;
  return raw >= 120 ? raw ~/ 60 : raw;
}

String weekdayShort(DateTime date) {
  const names = ['ПН', 'ВТ', 'СР', 'ЧТ', 'ПТ', 'СБ', 'ВС'];
  return names[date.weekday - 1];
}

const monthShort = ['янв.', 'фев.', 'мар.', 'апр.', 'мая', 'июн.', 'июл.', 'авг.', 'сен.', 'окт.', 'ноя.', 'дек.'];

String formDay(DateTime? at) {
  if (at == null) return '—';
  final d = at.toLocal();
  return '${d.day} ${monthShort[d.month - 1]}';
}

String recentGameLine({
  required String teamId,
  required String homeTeamId,
  required String awayTeamId,
  required int homeScore,
  required int awayScore,
  required String opponentName,
  DateTime? scheduledAt,
}) {
  final home = teamId == homeTeamId;
  final own = home ? homeScore : awayScore;
  final theirs = home ? awayScore : homeScore;
  return '$own:$theirs · $opponentName · ${formDay(scheduledAt)}';
}

const eventLabels = {
  'GOAL': 'Гол',
  'OWN_GOAL': 'Автогол',
  'ASSIST': 'Голевая',
  'YELLOW_CARD': 'Жёлтая',
  'RED_CARD': 'Красная',
  'SUBSTITUTION': 'Замена',
  'POINT': 'Очко',
  'FOUL': 'Фол',
  'PERIOD_START': 'Начало тайма',
  'PERIOD_END': 'Конец тайма',
  'OTHER': 'Событие',
};

String eventLabel(String? type) => eventLabels[type] ?? type ?? 'Событие';

String formatClock(int? totalSeconds) {
  final safe = totalSeconds == null || totalSeconds < 0 ? 0 : totalSeconds;
  final minutes = safe ~/ 60;
  final seconds = safe % 60;
  return '$minutes:${seconds.toString().padLeft(2, '0')}';
}

String periodLabel(int? period, {String? sportCode, int periodCount = 2}) {
  if (period == null || period < 1) return 'Ещё не свистнули';
  if (period > periodCount) return 'Доп. время';
  if (sportCode == 'BASKETBALL') return '$period-я четверть';
  if (sportCode == 'VOLLEYBALL') return '$period-я партия';
  return '$period-й тайм';
}

String playerTag(String? name, int? jersey) {
  if ((name == null || name.isEmpty) && jersey == null) return '';
  if (jersey == null) return name ?? '';
  return '#$jersey ${name ?? ''}'.trim();
}

String? matchOutcome({
  required String status,
  required String homeTeamId,
  required String awayTeamId,
  required int homeScore,
  required int awayScore,
  String? teamId,
}) {
  if (teamId == null || status != 'FINISHED') return null;
  final home = teamId == homeTeamId;
  if (!home && teamId != awayTeamId) return null;
  final scored = home ? homeScore : awayScore;
  final conceded = home ? awayScore : homeScore;
  if (scored > conceded) return 'WIN';
  if (scored < conceded) return 'LOSS';
  return 'DRAW';
}

const outcomeMark = {'WIN': 'В', 'DRAW': 'Н', 'LOSS': 'П'};
